README.md
# Kazarma

[![Latest Release](https://gitlab.com/technostructures/kazarma/kazarma/-/badges/release.svg)](https://gitlab.com/technostructures/kazarma/kazarma/-/releases)
[![Artifact Hub](https://img.shields.io/endpoint?url=https://artifacthub.io/badge/repository/technostructures)](https://artifacthub.io/packages/helm/technostructures/kazarma)
[![Matrix room](https://img.shields.io/matrix/kazarma:matrix.org)](https://matrix.to/#/#kazarma:matrix.org?via=matrix.org)
[![REUSE status](https://api.reuse.software/badge/gitlab.com/technostructures/kazarma/kazarma)](https://api.reuse.software/info/gitlab.com/technostructures/kazarma/kazarma)
[![Gitlab pipeline status](https://img.shields.io/gitlab/pipeline-status/kazarma/kazarma?branch=main)](https://gitlab.com/technostructures/kazarma/kazarma/-/pipelines)
[![coverage report](https://gitlab.com/technostructures/kazarma/kazarma/badges/main/coverage.svg)](https://gitlab.com/technostructures/kazarma/kazarma/-/commits/main)
[![translation status](https://hosted.weblate.org/widgets/kazarma/-/kazarma/svg-badge.svg)](https://hosted.weblate.org/engage/kazarma/)
[![Gitpod ready-to-code](https://img.shields.io/badge/Gitpod-ready--to--code-blue?logo=gitpod)](https://gitpod.io/#https://gitlab.com/technostructures/kazarma/kazarma)

A Matrix bridge to ActivityPub. It uses [this ActivityPub library](https://github.com/commonspub/ActivityPub) and [those](https://gitlab.com/technostructures/kazarma/matrix_app_service.ex) Matrix [libraries](https://gitlab.com/uhoreg/polyjuice_client).

## [User guide](https://docs.kazar.ma/category/user-guide)

![overview](doc_diagrams/overview.png)

## Installation

### [Erlang release](https://docs.kazar.ma/administrator-guide/erlang-release)

### [Docker Compose](https://docs.kazar.ma/administrator-guide/docker-compose)

### [Helm](https://docs.kazar.ma/administrator-guide/helm)

## Translation

Translation happens on [Hosted Weblate](https://hosted.weblate.org/engage/kazarma/).

[![translation status](https://hosted.weblate.org/widgets/kazarma/-/kazarma/horizontal-auto.svg)](https://hosted.weblate.org/engage/kazarma/)

## Development environment

### Using [Docker](https://docs.docker.com/engine/install/)

```bash
git clone git@gitlab.com:kazarma/kazarma.git
cd kazarma
git submodule update --init --recursive
docker compose run --rm synapse generate
docker compose run --rm kazarma mix do deps.get, ecto.setup
docker compose run --rm kazarma npm --prefix assets install
docker compose up
```

On Linux, use `docker-hoster` to make container domains accessible:
```
docker run -d \
    -v /var/run/docker.sock:/tmp/docker.sock \
    -v /etc/hosts:/tmp/hosts \
    --name docker-hoster \
    dvdarias/docker-hoster@sha256:2b0e0f8155446e55f965fa33691da828c1db50b24d5916d690b47439524291ba
```

(after rebooting, you will need to start it again using `docker start docker-hoster`)

This should run containers with those services:

- [kazarma.kazarma.local](http://kazarma.kazarma.local) -> Kazarma itself
- [matrix.kazarma.local](http://matrix.kazarma.local) -> Matrix server
- [kazarma.local](http://kazarma.local) -> serves .well-known routes that allow
  Matrix and Kazarma.ActivityPub to use simple `kazarma.local` domain (for
  users, etc)
- [pleroma.local](http://pleroma.local) -> Pleroma, should be able to address
  Matrix users using `kazarma.local` domain
- [element.local](http://element.local) -> Element, will connect to Synapse,
  should then be able to address Pleroma users using `pleroma.local` domain

You can also start an IEx console:

```bash
docker compose run --rm kazarma iex -S mix
```

#### On macOS

On macOS, instead of `docker-hoster` you need to add the following domains to your `/etc/hosts` file:
```
# Kazarma development domains
127.0.0.1 kazarma.local
127.0.0.1 kazarma.kazarma.local
127.0.0.1 matrix.kazarma.local
127.0.0.1 pleroma.local
127.0.0.1 element.local
```

Then the `docker-compose.yml` file should (at least) expose the `80` port in the `traefik` container:

```yaml
  traefik:
    image: traefik:v2.2.0
    ports:
      - 80:80
      - 443:443
```

#### Reset databases

```bash
docker compose rm -fs postgres_kazarma postgres_pleroma synapse
docker volume rm kazarma_postgres_pleroma_files kazarma_postgres_kazarma_files kazarma_synapse_files
docker compose run --rm synapse generate
docker compose run --rm kazarma mix ecto.setup
```

### Without using Docker

```bash
git clone git@gitlab.com:kazarma/kazarma.git
cd kazarma
git submodule update --init --recursive
mix do deps.get, ecto.setup
iex -S mix phx.server
```

## Run tests

```bash
mix test
```

## Generate documentation

We use [ditaa](http://ditaa.sourceforge.net) to generate diagrams and integrate them into HexDoc. To edit diagrams use [asciiflow](http://asciiflow.com/) and paste the result in HTML files in the `doc_diagrams` folder.

```bash
rm doc_diagrams/*.png && ditaa doc_diagrams/*.html
mix docs
```

## Sponsors

Kazarma receives funding from [NLnet](https://nlnet.nl/project/Kazarma-Release#ack) through the [NGI0 Entrust Fund](https://nlnet.nl/entrust/).

If you'd like to apply, check out how in [this video](https://media.ccc.de/v/36c3-10795-ngi_zero_a_treasure_trove_of_it_innovation)!

## License 

This project is available as open source under the terms of the AGPLv3. For accurate information, please check individual files.


CHANGELOG.md
# Changelog

## Unreleased

### Features

- Invite mentioned users in Notes in AP outbox rooms
- Invite mentioned users in Notes in direct message rooms

### Bug fixes

- Fix a bug in Webfinger request for relay actor

## [1.0.0-alpha.1]

### Features

#### Users and profiles

- Dynamic creation of Matrix puppets corresponding to ActivityPub Users
- Dynamic creation of ActivityPub puppets corresponding to Matrix Users
- Bidirectional update of display name and avatar

#### Generic bridging

- Mentions
- Deletions
- Attachments
- Replies
- Handling of activities: Post, Video and Event

#### Private messages

- Mastodon-like DMs using private (non-direct) rooms
- Pleroma-like chat using private (direct) rooms

#### Bridging of public activities from ActivityPub

- Opt-in using a follow of the relay actor, creates a public room
- Possibility of following by sending `!kazarma follow`
- Sending a message to the room automatically adds a mention
- Blocking from ActivityPub bans from the room
- Mentioning a Matrix user from ActivityPub invites them to the room if it exists

#### Bridging of a Matrix outbox to ActivityPub

- Creation of an outbox by inviting the kazarma bot and sending `!kazarma outbox`
- Possibility of following by sending `!kazarma follow`
- Sending a message to the room (by other users) automatically adds a mention

#### Specific ActivityPub implementations: Mobilizon groups

- Opt-in federation using instance follows
- Adding a Matrix user to a group creates a private group and invites them
- Matrix users accept the invitation by joining the room
- Removing them from the group bans them from the room
- The room is used for group discussions

#### Web front-end

- Search for a Matrix ID / AP ID / AP username
- Displays any user on ActivityPub or Matrix with their corresponding puppet on the other network
- Shows wether ActivityPub users have enabled public activity bridging
- Shows public activities from Matrix users that have activated their outbox room
- French and Spanish translations
- Integrated help

#### Deployment

- OCI image based on Alpine
- Helm chart
- Documentation on [https://docs.kazar.ma](https://docs.kazar.ma)


config/dev.exs
import Mix.Config
# SPDX-FileCopyrightText: 2020-2023 The Kazarma Team
# SPDX-License-Identifier: AGPL-3.0-only
# Configure your database
config :kazarma, Kazarma.Repo,
  username: System.get_env("POSTGRES_USER") || "postgres",
  password: System.get_env("POSTGRES_PASSWORD") || "postgres",
  database: System.get_env("POSTGRES_DB") || "kazarma_dev",
  hostname: System.get_env("POSTGRES_HOST") || "localhost",
  show_sensitive_data_on_connection_error: true,
  pool_size: 10

# For development, we disable any cache and enable
# debugging and code reloading.
#
# The watchers configuration can be used to run external
# watchers to your application. For example, we use it
# with webpack to recompile .js and .css sources.
config :kazarma, KazarmaWeb.Endpoint,
  http: [port: 4000],
  live_reload: [interval: 1000],
  # url: [host: "kazarma.kazarma.local", scheme: "https", port: 443],
  url: [host: "kazarma.kazarma.local", scheme: "http", port: 80],
  debug_errors: true,
  code_reloader: true,
  check_origin: false,
  watchers: [
    node: [
      "node_modules/webpack/bin/webpack.js",
      "--mode=development",
      "--watch",
      "--watch-options-stdin",
      cd: Path.expand("../assets", __DIR__)
    ]
  ]

# ## SSL Support
#
# In order to use HTTPS in development, a self-signed
# certificate can be generated by running the following
# Mix task:
#
#     mix phx.gen.cert
#
# Note that this task requires Erlang/OTP 20 or later.
# Run `mix help phx.gen.cert` for more information.
#
# The `http:` config above can be replaced with:
#
#     https: [
#       port: 4001,
#       cipher_suite: :strong,
#       keyfile: "priv/cert/selfsigned_key.pem",
#       certfile: "priv/cert/selfsigned.pem"
#     ],
#
# If desired, both `http:` and `https:` keys can be
# configured to run both http and https servers on
# different ports.

# Watch static and templates for browser reloading.
config :kazarma, KazarmaWeb.Endpoint,
  live_reload: [
    patterns: [
      ~r"priv/static/.*(js|css|png|jpeg|jpg|gif|svg)$",
      ~r"priv/gettext/.*(po)$",
      ~r"lib/phoenix_new_web/(live|views)/.*(ex)$",
      ~r"lib/phoenix_new_web/templates/.*(eex)$"
    ]
  ]

# Do not include metadata nor timestamps in development logs
config :logger, :console, format: "[$level] $message\n"

# Set a higher stacktrace during development. Avoid configuring such
# in production as building large stacktraces may be expensive.
config :phoenix, :stacktrace_depth, 20

# Initialize plugs at runtime for faster development compilation
config :phoenix, :plug_init_mode, :runtime

# MatrixAppService configuration
config :matrix_app_service, :app_service,
  base_url: "http://matrix.kazarma.local",
  access_token:
    "MDAyMGxvY2F0aW9uIG1hdHJpeC5pbWFnby5sb2NhbAowMDEzaWRlbnRpZmllciBrZXkKMDAxMGNpZCBnZW4gPSAxCjAwMmNjaWQgdXNlcl9pZCA9IEBhbGljZTptYXRyaXguaW1hZ28ubG9jYWwKMDAxNmNpZCB0eXBlID0gYWNjZXNzCjAwMjFjaWQgbm9uY2UgPSAjRC5ieWJrMEUuMU0qT0xJCjAwMmZzaWduYXR1cmUg-jGyjY9CK07mRt1h4p_86D6SJr1ZqrGr8YlsIW_jLtMK",
  homeserver_token:
    "MDAyMGxvY2F0aW9uIG1hdHJpeC5pbWFnby5sb2NhbAowMDEzaWRlbnRpZmllciBrZXkKMDAxMGNpZCBnZW4gPSAxCjAwMmNjaWQgdXNlcl9pZCA9IEBhbGljZTptYXRyaXguaW1hZ28ubG9jYWwKMDAxNmNpZCB0eXBlID0gYWNjZXNzCjAwMjFjaWQgbm9uY2UgPSBjcX4jazVTUDNeUlk2WnRECjAwMmZzaWduYXR1cmUg_K2biF-xm5ue7985RkAomVadF7yfy3UiEpH-e15m0esK"

config :matrix_app_service, :client_options,
  ssl_options: [
    verify: :verify_none
  ]

config :activity_pub, :domain, "kazarma.local"

config :activity_pub, :http,
  adapter: [
    ssl_options: [
      verify: :verify_none
    ]
  ]

config :kazarma, bridge_remote_matrix_users: true
config :kazarma, html_search: true
config :kazarma, html_actor_view_include_remote: true
config :kazarma, frontpage_help: true
config :kazarma, frontpage_before_text: nil
config :kazarma, frontpage_after_text: nil

config :matrix_app_service, ignore_exceptions: true
